src/tls_stream.rs

```diff
@@ -541,17 +545,27 @@ impl<S> TlsStream<S>
     }
 
     fn read_in(&mut self) -> io::Result<usize> {
-        let existing_len = self.enc_in.position() as usize;
-        let min_len = cmp::max(1024, 2 * existing_len);
-        if self.enc_in.get_ref().len() < min_len {
-            self.enc_in.get_mut().resize(min_len, 0);
+        let mut sum_nread = 0;
+
+        while self.needs_read > 0 {
+            let existing_len = self.enc_in.position() as usize;
+            let min_len = cmp::max(1024, self.needs_read + 2 * existing_len);
```

Comment:
This seems a little suspicious - maybe it should be something like `max(max(1024, existing_len * 2), self.needs_read)`? We want to make sure our buffer is larger than 1023, larger than the amount we want to read, and is doubling in size past that point.